### DISCUSSION TITLE
KeyBindings take priority over control key handler

### DISCUSSION THREAD

**Discussion by ClxS** [AvaloniaUI/Avalonia#15771] [Q&A] [answered]
I'm running into an issue where my keybinding gestures are taking priority over a textbox input.
My situation:
Picture a hierarchical treeview such as the Rider solution explorer. I have a KeyBinding setup on the root of this for the Delete key. This binding will cause all the selected nodes to be deleted. (as a single transaction, hence the root keybinding)
I have also added an F2 action which will place a textbox into the tree, allowing you to rename the element.

Pressing the Delete key while editing text in this TextBox will delete the node.
I have tried capturing the input and marking it as handled, the problem is that it appears that the KeyBindings are _always_ evaluated first and RaiseEvent on the element afterwards. This makes it so the element handling the input doesn't matter, the delete action will always run.
That is handled in the KeyboardDevice here: https://github.com/AvaloniaUI/Avalonia/blob/1b27bf7bac95872b7934337fb181fcb0ebbfa797/src/Avalonia.Base/Input/KeyboardDevice.cs#L204

Does anyone have any ideas on how to workaround this? I'm currently looking at just replacing the entire KeyboardDevice to instead call RaiseEvent first and skip keybindings if the event is handled, but I'd prefer not to do this if I could avoid it.

Edit: Looks like it's impossible to replace the KeyboardDevice because the singleton accessor KeyboardDevice.Instance used in some places tries to cast.

**Comment by ClxS** [comment]
Unfortunately my plan to just replace the KeyboardDevice locally doesn't work, as places in Avalonia code use KeyboardDevice.Instance which casts to KeyboardDevice.

I have come up with a solution which is to add keybindings on my textboxes themselves to supress the problem keys (Delete, Arrow Keys), which has the side effect that the default behaviour for those keys doesn't work so I've had to wire it up via reflection. Messy, but it works.

**Comment by renrutsirhc** [comment]
This sounds really similar to a problem I had using the delete key to remove listboxitems that contained textboxes. My workaround was to use got focus and lost focus events for the textbox to set a property on the vm - something like isEditing. You can use this property to disable the command you bind to the delete keybinding. With the keybinding command disabled, the events propagate as normal.

Interested if there are other solutions for this out there!

**Comment by timunie** [comment]
Have you tunneled the event?

**Reply by ClxS** [reply]
No, the KeyBinding system does not support any configurable routing. They work _before_ and outside of the ordinary event bubbling system which is the root cause of the problem here.

The flow for what happens on a focused keyevent is
1) From the target visual element, bubble up to the root and check if any matching key bindings exist. If they do and a command is set for that binding, set the key event Handled to true
2) Build the EventRoute and raise the event.

There's three issues here:
1) Because the KeyBindings system runs separately and before the EventRoute system, it effectively means the KeyBindings system has absolute priority. It's impossible to block that key binding via tunnelling because the bubbled up KeyBindings will have ran first.
2) A matching keybinding even existing anywhere in the ancestor tree sets handled to true, which then stops events being raised along the event route.
3) KeyBinding commands are simple and have no context about the origin of the underlying keyevent, making it not possible for me to shove a hack into my commands to stop this behaviour by disabling the command if the source target was a textbox.

Ultimately I feel this isn't so much a bug, but a design flaw in how KeyBindings are implemented. Shouldn't the keybinding check occur at the same time as the bubbling pass, and respect Handled at that point?

**Comment by BAndysc** [comment]
The problem is not that KeyBindings are evaluated first, the main problem is that TextBox doesn't use KeyBindings.

I feel like every developer has this issue. This is my workaround: https://github.com/BAndysc/WoWDatabaseEditor/blob/ad03280fc13e3544e46d0b1ace550973c72ef4f6/WDE.Common.Avalonia/Controls/BetterKeyBinding.cs#L15

I replaced the default KeyBinding with my subclass which wraps the command by checking if the focus is in a TextBox. In this case, it doesn't execute the command. Far from perfect, but works for me.

**Reply by ClxS** [reply]
I've no doubt that the "BetterKeyBinding" will work for this specific case, but I still argue this is a design flaw in the KeyBinding system. The problem with a replacement KeyBinding like that is you'd have to constantly extend it as you have more controls which want to handle keys. To make it properly universal without having a hard coded list of types you're essentially rewriting a poor mans version of RaiseEvent method, while losing support for Tunneling.

To make the BetterKeyBinding work in a robust way, I would have to create a keybinding for every key a textbox could handle and add that as a gesture to every textbox. That quickly gets out of control and would result in tens of thousands of KeyBinding instances being created in a large app. This would be required because it's common to support singular keys bindings like "F" to focus an object in a lot of applications.

**Comment by ClxS** [comment, marked as answer]
I've come up with this, it's as close as I can get:
```cs

public class FixKeyBindings : Behavior<Control>
{
    private KeyBinding[] keyBindings;
    
    public static readonly StyledProperty<IEnumerable<Key>> KeysProperty = AvaloniaProperty.Register<FixKeyBindings , IEnumerable<Key>>("Keys");

    public FixKeyBindings()
    {
        this.keyBindings = Keys
            .Select(
                key =>
                {
                    return new KeyBinding()
                    {
                        Gesture = new KeyGesture(key),
                        Command = ReactiveCommand.Create(() => { }),
                    };
                })
            .ToArray();
    }
    
    public IEnumerable<Key> Keys
    {
        get => GetValue(KeysProperty);
        set => SetValue(KeysProperty, value);
    }
    
    protected override void OnAttached()
    {
        base.OnAttached();

        if (this.AssociatedObject is not null)
        {
            this.AssociatedObject.AddHandler(InputElement.KeyDownEvent, FlagHandledKeys, handledEventsToo: true, routes: RoutingStrategies.Tunnel);
            this.AssociatedObject.KeyBindings.AddRange(this.keyBindings);
        }
    }

    protected override void OnDetaching()
    {
        if (this.AssociatedObject is not null)
        {
            this.AssociatedObject.RemoveHandler(InputElement.KeyDownEvent, FlagHandledKeys);
            this.AssociatedObject.KeyBindings.RemoveMany(this.keyBindings);
        }
        
        base.OnDetaching();
    }
    
    private void FlagHandledKeys(object? s, KeyEventArgs e)
    {
        if (Keys.Contains(e.Key))
        {
            e.Handled = false;
        }
    }
}
```

How this works:
1) Registers a KeyBinding on the textbox itself for the problem keys. This binding exists solely to block the bindings higher up the tree.
2) Registers a tunnelled keydown handler who's purpose is to unblock the e.Handled, which was set due to the KeyBinding.
3) The ordinary Bubble flow will work for the textbox now, input works, and higher up KeyBindings are blocked.
